#10923, #10924: Disabling GeoFence Rules and Add filter by IP for GeoFence rules

mahmoudadel54 opened the pull request:
## Description
<!-- A few sentences describing the overall goals of the pull request' s commits. -->
This PR includes:
- adding IPAddress filter in rules manage grid
- adding Date filter  in rules manage grid for validity period fields
- add validity period fields for rule insertion/update

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue
#10923 #10924 

**What is the current behavior?**
#10923 #10924 

**What is the new behavior?**
IPAddress filter is added to rules manager grid and date filter as well.
Now to add rule user can enter validity period fields 'validAfter', 'validBefore' in insert new rule or update existing rule

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information
* https://docs.geoserver.org/latest/en/user/extensions/geofence-server/rest.html#filter-parameters
* https://docs.geoserver.org/latest/en/user/extensions/geofence-server/rest.html#rule-content
* https://tehnoblog.org/ip-tools/ip-address-in-cidr-range/

Testing:

Can be tested on nightly geoserver/geofence.

GeoFence:

https://build.geoserver.org/geofence/nightly/main/geofence-main-latest-war.zip

GeoServer

https://build.geoserver.org/geoserver/main/geoserver-main-latest-war.zip
https://build.geoserver.org/geoserver/main/ext-latest/geoserver-2.27-SNAPSHOT-geofence-server-h2-plugin.zip

offtherailz commented:
@tdipisa by they way.
I noticed that the documentation for the pages 
https://mapstore.geosolutionsgroup.com/mapstore/docs/api/plugins#pages.RulesManager

Disappeared on dev 

https://dev-mapstore.geosolutionsgroup.com/mapstore/docs/api/plugins#pages.RulesManager

Trying to apply tthe same config, I have this result

![image](https://github.com/user-attachments/assets/4f7a49f1-646e-4343-bc7d-f30e8755963c)
While I see on dev it works well: 
![image](https://github.com/user-attachments/assets/69ce62f2-ea23-40e8-b144-6e0513417c7d)

So we have to update the doc properly to explain how to setup rule-manager.

offtherailz reviewed (changes requested):
Testing this PR widely is very hard by per-se, because we should need to test many use cases:
- Standalone new version (nightly)
- Embedded new version (nightly)
- Standalone old version (I suppose this tool needs to work with current and old versions of geoserver/geofence too)
- Embedded old version (I suppose this tool needs to work with current and old versions of geoserver/geofence too)

Moreover configuring everything is not easy. I had the opporunity, in the time I had dedicated on it, to test only the standalone version.
I had a lot of crashes to report, so I stopped reviewing after reporting 3 different crashes, waiting for a more stable version.

Testing all the version should not be up to me, but we could provide a structure to switch and test easely the various version to testers/develoeprs.

- [ ] See the comment for configuration and documentation above.
- [ ] Because geofence improvemnts are still on nightly, we have to ensure that it will work also for existing version of geoserver with compatibility, I didn't test it, but we have to proof this change will not be distruptive. Maybe add an opt-in ?
- [ ] I noticed this bug, not sure it is related to the PR, but on dev is not reroducible.

https://github.com/user-attachments/assets/832c6211-2e80-4288-a1c8-fda8440a5240

```
StandardRouter.jsx:88 TypeError: Cannot read properties of null (reading 'scrollTop')
    at DataGrid.eval [as scrollListener] (DataGrid.jsx:29:35)
    at RulesGrid.eval [as componentDidUpdate] (RulesGrid.jsx:46:19)
```
- [ ] The `ipAddressAny` parameter do not have any effect. I'd expect it to filter by precise IP (range) vs if the rule is applied. Verify if it makes sense, or if it doesn't show it checked and disable. The same looks to be 


https://github.com/user-attachments/assets/fd4e5ecc-aa4d-44e4-aa6b-67308f1e6767



- [ ] See the video above (for ipAddressAny, I had a crash while recording .

```
StandardRouter.jsx:88 TypeError: Cannot read properties of undefined (reading 'id')
    at eval (react-data-grid-addons.js:9672:39)
    at Array.filter (<anonymous>)
    at CustomDragLayer.isDraggedRowSelected (react-data-grid-addons.js:9671:30)
    at CustomDragLayer.getDraggedRows (react-data-grid-addons.js:9690:29)
    at CustomDragLayer.renderDraggedRows (react-data-grid-addons.js:9702:20)
    at CustomDragLayer.render (react-data-grid-addons.js:9745:31)
    at finishClassComponent (react-dom.development.js:18407:31)
    at updateClassComponent (react-dom.development.js:18360:24)
    at beginWork$1 (react-dom.development.js:20108:16)
    at HTMLUnknownElement.callCallback (react-dom.development.js:362:14)
```

- [ ] Another crash while playing on date filter. 

https://github.com/user-attachments/assets/45614bf1-bc31-481e-a4e8-11dad6d67ce4


```
StandardRouter.jsx:88 TypeError: Cannot destructure 'state.filters' as it is undefined.
    at rulesmanager (rulesmanager.js:161:15)
    at combination (combineReducers.js:126:29)
    at reduce (StateUtils.js:256:14)
    at standardRootReducerFunc (defaultOptions.js:79:8)
    at rootReducer (StandardStore.js:90:12)
```

tdipisa commented:
@offtherailz 

>@tdipisa by they way.
>I noticed that the documentation for the pages
>https://mapstore.geosolutionsgroup.com/mapstore/docs/api/plugins#pages.RulesManager

>Disappeared on dev

>https://dev-mapstore.geosolutionsgroup.com/mapstore/docs/api/plugins#pages.RulesManager

thank you for informing about that. I think we have to check why and restore, then. Please take care of this as part of this PR @mahmoudadel54

mahmoudadel54 commented:
- yeah, the 1st issue is a bug not related to the work in this issue [I fixed it]
- for the ipAddressAny parameter do not have any effect. you should use this link https://tehnoblog.org/ip-tools/ip-address-in-cidr-range/ which I added into the PR description above to validate your filter
- for the crashing in ipAddressAny filter please give me clear steps to reproduce as I have tried many time but can't reproduce it 
- for the date filter by writing date manually --> I will fix it 



> * [ ]  I noticed this bug, not sure it is related to the PR, but on dev is not reroducible.
> 
>  MapStore.mp4 
> ```
> StandardRouter.jsx:88 TypeError: Cannot read properties of null (reading 'scrollTop')
>     at DataGrid.eval [as scrollListener] (DataGrid.jsx:29:35)
>     at RulesGrid.eval [as componentDidUpdate] (RulesGrid.jsx:46:19)
> ```
> 
> * [ ]  The `ipAddressAny` parameter do not have any effect. I'd expect it to filter by precise IP (range) vs if the rule is applied. Verify if it makes sense, or if it doesn't show it checked and disable. The same looks to be
> 
>  MapStore.1.mp4 
> * [ ]  See the video above (for ipAddressAny, I had a crash while recording .
> 
> ```
> StandardRouter.jsx:88 TypeError: Cannot read properties of undefined (reading 'id')
>     at eval (react-data-grid-addons.js:9672:39)
>     at Array.filter (<anonymous>)
>     at CustomDragLayer.isDraggedRowSelected (react-data-grid-addons.js:9671:30)
>     at CustomDragLayer.getDraggedRows (react-data-grid-addons.js:9690:29)
>     at CustomDragLayer.renderDraggedRows (react-data-grid-addons.js:9702:20)
>     at CustomDragLayer.render (react-data-grid-addons.js:9745:31)
>     at finishClassComponent (react-dom.development.js:18407:31)
>     at updateClassComponent (react-dom.development.js:18360:24)
>     at beginWork$1 (react-dom.development.js:20108:16)
>     at HTMLUnknownElement.callCallback (react-dom.development.js:362:14)
> ```
> 
> * [ ]  Another crash while playing on date filter.
> 
>  MapStore.2.mp4 
> ```
> StandardRouter.jsx:88 TypeError: Cannot destructure 'state.filters' as it is undefined.
>     at rulesmanager (rulesmanager.js:161:15)
>     at combination (combineReducers.js:126:29)
>     at reduce (StateUtils.js:256:14)
>     at standardRootReducerFunc (defaultOptions.js:79:8)
>     at rootReducer (StandardStore.js:90:12)
> ```

allyoucanmap commented:
@tdipisa @offtherailz after the changes and clean up of JSDoc (#10814) that includes the automated script to list plugins now only plugins will checked and go in the plugins doc. The RulesManager is a page and not a plugin so afterward I moved it in the framework doc (#10877) so in the new doc for this page is here https://dev-mapstore.geosolutionsgroup.com/mapstore/docs/api/framework#pages.RulesManager. If we want to move it again inside the plugins doc we need to improve a bit the [jsDocUtils.js](https://github.com/geosolutions-it/MapStore2/blob/master/utility/doc/jsDocUtils.js) to include pages automatically inside the plugins documentation

offtherailz commented:
So @mahmoudadel54 for the moment linking (with a relative path) the plugin to the page doc is enough. Something like: "This Plugin requires the configuration of the rules-manager page (link) to be activated. "

offtherailz reviewed (approved):
LGTM. I didn't tested on GeoServer integrated version, but it can be tested as well as deployed on dev.
There is only 1 bug I noticed that have to be fixed ON BACKEND. @etj 
- [ ] There is a bug on filtering checkboxes. It happens only on IP and Date columns. Here in the video I show the IP's one, but the same can be done on date column

https://github.com/user-attachments/assets/807e5f50-f4f0-47ff-9561-49a908bdb635

The backend replies on dateAny params true or false in the same way, but for `ipAddressAny` or `dateAny` = true, it doesn't return any value. For the other parameters (e.g. `requestAny` = true) it simply ignores the param and works by returning all the values (because we don't have the corrispondent parameter to check). 

![image](https://github.com/user-attachments/assets/bf8235ed-538d-4b63-bb38-6a9b92cea02f)

basically this do not return anything:
```
http://localhost:8081/geofence/rest/rules/count?dateAny=true&layerAny=true&ipAddressAny=true&requestAny=true
```

and this returns everything

```
http://localhost:8081/geofence/rest/rules/count?layerAny=true&requestAny=true
```

That is unconsistent.


- [ ] the `date` filter without `dateAny` do not have any difference. In this video you can see that I should be able to search "all dates that match the criteria" vs search by criteria. But both the request look only to work to match the criteria.

https://github.com/user-attachments/assets/c0d5ee74-e01a-4260-a641-dcab76a81686

offtherailz commented:
@etj  I was reading the actual tooltips for the checkbox to deduce functionality. In fact I think that while for the other rules this can be the same, in this case setting the velue Any to true has not the same meaning (because of range nature of these fields).
Maybe we have to change the tooltip, hide the checkboxes for these fields, or properly implement the functionality on BE.
@tdipisa let's check this point asap.

- ![image](https://github.com/user-attachments/assets/43b058a5-04ad-4447-86f3-ea8e02142651)

- ![image](https://github.com/user-attachments/assets/0a61b1dc-8ec3-4c23-91e4-991fc384e83a)

tdipisa commented:
@offtherailz and @etj thank you so much for clarifying the above.

@mahmoudadel54 we need you now to provide:
- The fix to the filter check logic
- The tooltip improvement

see [above](https://github.com/geosolutions-it/MapStore2/pull/10925#issuecomment-2760916825) for more details.

offtherailz reviewed (approved):
@tdipisa I had to fix tooltips initially achieved because translation where a lot misleading. 
Can you take a look to have a second opinion. Thank you.

(I also fixed tootip position that was making the tooltip go out of screen for some fields.

tdipisa commented on web/client/translations/data.it-IT.json:
```diff
@@ -2848,8 +2850,9 @@
         "close": "Esci da modifica regola",
-        "showAllRules": "Mostra tutte le regole applicabili",
-        "filterRuleList": "Filtra la lista usando il valore selezionato"
+        "showAllRules": "Mostra solo le regole che corrispondono specificamente in base al valore di questo campo. Deseleziona per includere anche le regole dove questo campo non è considerato.",
```
```suggestion
        "showAllRules": "Mostra solo le regole che corrispondono strettamente al valore indicato per questo campo. Deseleziona per includere anche le regole dove tale valore non è presente.",
```

tdipisa commented on web/client/translations/data.it-IT.json:
```diff
@@ -2848,8 +2850,9 @@
-        "showAllRules": "Mostra tutte le regole applicabili",
-        "filterRuleList": "Filtra la lista usando il valore selezionato"
+        "showAllRules": "Mostra solo le regole che corrispondono specificamente in base al valore di questo campo. Deseleziona per includere anche le regole dove questo campo non è considerato.",
+        "filterRuleList": "Mostra tutte le regole applicate, anche se non corrispondono specificamente in base a questo campo. Seleziona per visualizzare solo le regole che si applicano a causa di questo valore.",
```
```suggestion
        "filterRuleList": "Mostra solo le regole dove il valore specificato non è presente. Deseleziona per includere solo le regole che corrispondono strettamente al valore indicato per questo campo.",
```

offtherailz replied:
This is incorrect for 2 reasons: 
- The tooltip is shown when the item is unchecked, so the user have to (check), not "uncheck" (deseleziona)
- In the condition where the tooltip is shown, the table show all rules that match, including both request that match and that are empty. Here your text says to show "only" unselected.
```suggestion
        "filterRuleList": "Attualmente la tabella mostra tutte le regole applicate per questa condizione, anche quelle per cui questo valore non è impostato . Seleziona per visualizzare solo le regole che si applicano specificatamente per questo valore.",
```

tdipisa reviewed (changes requested):
@offtherailz I would write like this (see my change). If you agreed, please translate the same for other languages.

offtherailz commented on web/client/translations/data.it-IT.json:
```diff
@@ -2848,8 +2850,9 @@
         "close": "Esci da modifica regola",
-        "showAllRules": "Mostra tutte le regole applicabili",
-        "filterRuleList": "Filtra la lista usando il valore selezionato"
+        "showAllRules": "Mostra solo le regole che corrispondono strettamente al valore indicato per questo campo. Deseleziona per includere anche le regole dove tale valore non è presente.",
```
```suggestion
        "showAllRules": "Attualmente la tabella mostra solo le regole che corrispondono strettamente al valore indicato per questo campo. Deseleziona per includere anche le regole dove tale valore non è presente.",
```

tdipisa commented:
@ElenaGallo can you please test this and let me know if we can backport to 2025.01.xx?

offtherailz commented:
@ElenaGallo, could you please test this on  DEV  ? Thank you
Ideally we should test:

- Standalone new version (nightly)
- Embedded new version (nightly)
- Standalone old version (I suppose this tool needs to work with current and old versions of geoserver/geofence too)
- Embedded old version (I suppose this tool needs to work with current and old versions of geoserver/geofence too)

Actually on dev we can test only embedded old version. Let's check if the functionalitries break something (e.g. adding an ip rule or a date rule, trying to filter by ip or by date).

Try to save edit and filter also by other fields (e.g. user or layer), to verify the functionalities are preserved.

ElenaGallo commented:
@mahmoudadel54 @tdipisa The Date functionality works, but the Date filter is not clear to me. 
If I enter the interval >2025-11-18 & <2025-11-19 in the rule, when I filter for the date 2025-11-18 in the filter box, the rule with >2025-11-18 & <2025-11-19 does not appear in the list.
In this case, shouldn’t the date 2025-11-18 be included?

![date](https://github.com/user-attachments/assets/e6d4c8ca-b93a-4c44-8e42-37882e3399b3)
